File: app/src/main/java/org/wikipedia/history/HistoryEntry.java
package org.wikipedia.history;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.apache.commons.lang3.StringUtils;
import org.wikipedia.page.PageTitle;

import java.util.Date;

public class HistoryEntry implements Parcelable {
    public static final HistoryEntryDatabaseTable DATABASE_TABLE = new HistoryEntryDatabaseTable();

    public static final int SOURCE_SEARCH = 1;
    public static final int SOURCE_INTERNAL_LINK = 2;
    public static final int SOURCE_EXTERNAL_LINK = 3;
    public static final int SOURCE_HISTORY = 4;
    public static final int SOURCE_LANGUAGE_LINK = 6;
    public static final int SOURCE_RANDOM = 7;
    public static final int SOURCE_MAIN_PAGE = 8;
    public static final int SOURCE_DISAMBIG = 10;
    public static final int SOURCE_READING_LIST = 11;
    public static final int SOURCE_FEED_CONTINUE_READING = 12;
    public static final int SOURCE_FEED_BECAUSE_YOU_READ = 13;
    public static final int SOURCE_FEED_MOST_READ = 14;
    public static final int SOURCE_FEED_FEATURED = 15;
    public static final int SOURCE_NEWS = 16;
    public static final int SOURCE_FEED_MAIN_PAGE = 17;
    public static final int SOURCE_FEED_RANDOM = 18;
    public static final int SOURCE_GALLERY = 19; // Or should we just use SOURCE_INTERNAL_LINK as before?  Some of these things seem not like the others.
    public static final int SOURCE_APP_SHORTCUT_RANDOM = 20;
    public static final int SOURCE_APP_SHORTCUT_CONTINUE_READING = 21;
    public static final int SOURCE_FEED_MOST_READ_ACTIVITY = 22;
    public static final int SOURCE_ON_THIS_DAY_CARD = 23;
    public static final int SOURCE_ON_THIS_DAY_ACTIVITY = 24;
    public static final int SOURCE_NOTIFICATION = 25;
    public static final int SOURCE_NOTIFICATION_SYSTEM = 26;
    public static final int SOURCE_FLOATING_QUEUE = 27;
    public static final int SOURCE_EDIT_DESCRIPTION = 28;
    public static final int SOURCE_WIDGET = 29;
    public static final int SOURCE_SUGGESTED_EDITS = 30;
    public static final int SOURCE_TALK_TOPIC = 31;
    public static final int SOURCE_WATCHLIST = 32;
    public static final int SOURCE_EDIT_DIFF_DETAILS = 33;
    public static final int SOURCE_ERROR = 34;

    @NonNull private final PageTitle title;
    @NonNull private final Date timestamp;
    private final int source;
    private final int timeSpentSec;

    // Transient variable, not stored in the db, to be set when navigating back and forth between articles.
    @Nullable private String referrer;

    public HistoryEntry(@NonNull PageTitle title, @NonNull Date timestamp, int source,
                        int timeSpentSec) {
        this.title = title;
        this.timestamp = timestamp;
        this.source = source;
        this.timeSpentSec = timeSpentSec;
    }

    public HistoryEntry(@NonNull PageTitle title, @NonNull Date timestamp, int source) {
        this(title, timestamp, source, 0);
    }

    public HistoryEntry(@NonNull PageTitle title, int source) {
        this(title, new Date(), source);
    }

    @NonNull public PageTitle getTitle() {
        return title;
    }

    @NonNull public Date getTimestamp() {
        return timestamp;
    }

    public int getSource() {
        return source;
    }

    public int getTimeSpentSec() {
        return timeSpentSec;
    }

    public void setReferrer(@Nullable String referrer) {
        this.referrer = referrer;
    }

    @Nullable public String getReferrer() {
        return referrer;
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof HistoryEntry)) {
            return false;
        }
        HistoryEntry other = (HistoryEntry) o;
        return title.equals(other.title)
                && timestamp.equals(other.timestamp)
                && source == other.source;
    }

    @Override
    public int hashCode() {
        int result = title.hashCode();
        result = 31 * result + source;
        result = 31 * result + timestamp.hashCode();
        result = 31 * result + timeSpentSec;
        return result;
    }

    @Override
    public String toString() {
        return "HistoryEntry{"
                + "title=" + title
                + ", source=" + source
                + ", timestamp=" + timestamp.getTime()
                + ", timeSpentSec=" + timeSpentSec
                + '}';
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeParcelable(getTitle(), flags);
        dest.writeLong(getTimestamp().getTime());
        dest.writeInt(getSource());
        dest.writeInt(getTimeSpentSec());
        dest.writeString(StringUtils.defaultString(referrer));
    }

    private HistoryEntry(Parcel in) {
        this.title = in.readParcelable(PageTitle.class.getClassLoader());
        this.timestamp = new Date(in.readLong());
        this.source = in.readInt();
        this.timeSpentSec = in.readInt();
        this.referrer = in.readString();
    }

    public static final Parcelable.Creator<HistoryEntry> CREATOR
            = new Parcelable.Creator<HistoryEntry>() {
        @Override
        public HistoryEntry createFromParcel(Parcel in) {
            return new HistoryEntry(in);
        }

        @Override
        public HistoryEntry[] newArray(int size) {
            return new HistoryEntry[size];
        }
    };
}

File: app/src/main/java/org/wikipedia/history/HistoryEntry.kt
package org.wikipedia.history

import android.os.Parcelable
import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.PrimaryKey
import kotlinx.parcelize.IgnoredOnParcel
import kotlinx.parcelize.Parcelize
import kotlinx.parcelize.TypeParceler
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient
import org.wikipedia.dataclient.WikiSite
import org.wikipedia.json.DateSerializer
import org.wikipedia.page.PageTitle
import org.wikipedia.parcel.DateParceler
import java.util.*

@Serializable
@Parcelize
@TypeParceler<Date, DateParceler>()
@Entity
// TODO: change these properties back to val when HistoryEntry is no longer serializable. (i.e. when we update Tabs to be in the database instead of Prefs)
class HistoryEntry(
    var authority: String = "",
    var lang: String = "",
    var apiTitle: String = "",
    var displayTitle: String = "",
    @PrimaryKey(autoGenerate = true) var id: Int = 0,
    var namespace: String = "",
    @Serializable(with = DateSerializer::class) var timestamp: Date = Date(),
    var source: Int = SOURCE_INTERNAL_LINK,
    var timeSpentSec: Int = 0,
) : Parcelable {
    constructor(title: PageTitle, source: Int, timestamp: Date = Date(), timeSpentSec: Int = 0) : this(title.wikiSite.authority(),
        title.wikiSite.languageCode, title.text, title.displayText, namespace = title.namespace,
        timestamp = timestamp, source = source, timeSpentSec = timeSpentSec) {
        pageTitle = title
    }

    @IgnoredOnParcel
    @Transient
    @Ignore
    private var pageTitle: PageTitle? = null

    val title: PageTitle get() {
        if (pageTitle == null) {
            pageTitle = PageTitle(namespace, apiTitle, WikiSite(authority, lang))
            pageTitle!!.displayText = displayTitle
        }
        return pageTitle!!
    }

    // To be set when navigating back and forth between articles.
    @IgnoredOnParcel
    @Transient
    @Ignore
    var referrer: String? = null

    companion object {
        const val SOURCE_SEARCH = 1
        const val SOURCE_INTERNAL_LINK = 2
        const val SOURCE_EXTERNAL_LINK = 3
        const val SOURCE_HISTORY = 4
        const val SOURCE_LANGUAGE_LINK = 6
        const val SOURCE_RANDOM = 7
        const val SOURCE_MAIN_PAGE = 8
        const val SOURCE_DISAMBIG = 10
        const val SOURCE_READING_LIST = 11
        const val SOURCE_FEED_CONTINUE_READING = 12
        const val SOURCE_FEED_BECAUSE_YOU_READ = 13
        const val SOURCE_FEED_MOST_READ = 14
        const val SOURCE_FEED_FEATURED = 15
        const val SOURCE_NEWS = 16
        const val SOURCE_FEED_MAIN_PAGE = 17
        const val SOURCE_FEED_RANDOM = 18
        const val SOURCE_GALLERY = 19
        const val SOURCE_APP_SHORTCUT_RANDOM = 20
        const val SOURCE_APP_SHORTCUT_CONTINUE_READING = 21
        const val SOURCE_FEED_MOST_READ_ACTIVITY = 22
        const val SOURCE_ON_THIS_DAY_CARD = 23
        const val SOURCE_ON_THIS_DAY_ACTIVITY = 24
        const val SOURCE_NOTIFICATION = 25
        const val SOURCE_NOTIFICATION_SYSTEM = 26
        const val SOURCE_FLOATING_QUEUE = 27
        const val SOURCE_EDIT_DESCRIPTION = 28
        const val SOURCE_WIDGET = 29
        const val SOURCE_SUGGESTED_EDITS = 30
        const val SOURCE_TALK_TOPIC = 31
        const val SOURCE_WATCHLIST = 32
        const val SOURCE_EDIT_DIFF_DETAILS = 33
        const val SOURCE_ERROR = 34
        const val SOURCE_EDIT_HISTORY = 35
        const val SOURCE_CATEGORY = 36
    }
}
